Invoke-SqlCmd2/Public/Invoke-SqlCmd2.ps1

```diff
@@ -315,225 +322,236 @@ function Invoke-Sqlcmd2 {
+            }
+            else {
+                throw "SQLConnection is not open"
+            }
+        }
+        $GoSplitterRegex = [regex]'(?smi)^[\s]*GO[\s]*$'
+
+    }
+    process {
+        foreach ($SQLInstance in $ServerInstance) {
+            Write-Verbose "Querying ServerInstance '$SQLInstance'"
+
+            if ($PSBoundParameters.Keys -contains "SQLConnection") {
+                $Conn = $SQLConnection
+            }
+            else {
+                $CSBuilder = New-Object -TypeName System.Data.SqlClient.SqlConnectionStringBuilder
+                $CSBuilder["Server"] = $SQLInstance
+                $CSBuilder["Database"] = $Database
+                $CSBuilder["Connection Timeout"] = $ConnectionTimeout
+
+                if ($Encrypt) {
+                    $CSBuilder["Encrypt"] = $true
+                }
+
+                if ($Credential) {
+                    $CSBuilder["Trusted_Connection"] = $false
+                    $CSBuilder["User ID"] = $Credential.UserName
+                    $CSBuilder["Password"] = $Credential.GetNetworkCredential().Password
+                }
+                else {
+                    $CSBuilder["Integrated Security"] = $true
+                }
+                if ($ApplicationName) {
+                    $CSBuilder["Application Name"] = $ApplicationName
+                }
+                else {
+                    $ScriptName = (Get-PSCallStack)[-1].Command.ToString()
+                    if ($ScriptName -ne "<ScriptBlock>") {
+                        $CSBuilder["Application Name"] = $ScriptName
+                    }
+                }
+                $conn = New-Object -TypeName System.Data.SqlClient.SQLConnection
+
+                $ConnectionString = $CSBuilder.ToString()
+                $conn.ConnectionString = $ConnectionString
+                Write-Debug "ConnectionString $ConnectionString"
+
+                try {
+                    $conn.Open()
+                }
+                catch {
+                    Write-Error $_
+                    continue
+                }
+            }
+
+            #Following EventHandler is used for PRINT and RAISERROR T-SQL statements. Executed when -Verbose parameter specified by caller
+            if ($PSBoundParameters.Verbose) {
+                $conn.FireInfoMessageEventOnUserErrors = $false # Shiyang, $true will change the SQL exception to information
+                $handler = [System.Data.SqlClient.SqlInfoMessageEventHandler] { Write-Verbose "$($_)" }
+                $conn.add_InfoMessage($handler)
+            }
+            if ($ParseGO) {
+                Write-Verbose "Stripping GOs from source"
+                $Pieces = $GoSplitterRegex.Split($Query)
+            }
+            else {
+                $Pieces = , $Query
+            }
+            # Only execute non-empty statements
+            $Pieces = $Pieces | Where-Object { $_.Trim().Length -gt 0 }
```

Comment:
If there's a chance `$null` could make it into the `$Pieces` collection, might be worth checking for null here, or casting pieces as `[string[]]` when declared above.  Otherwise, could get errors re: trying trim method on null

Cheers!

Reply:
doh, you're correct.

Reply:
nah, wait. "$pieces" comes from a split. it can be `-eq ''` but not `$null`